tests/test_dynamo_db_block.py

```diff
@@ -0,0 +1,213 @@
+        blk = SaveCounterDynamoDB()
+        self.configure_block(blk, {
+            'table': '{{$table}}',
+            'log_level': 'DEBUG'
+        })
+        blk.start()
+
+        blk.process_signals([Signal({
+            'table': 'odd' if i % 2 else 'even',
+            'val': i,
+            '_id': ''
+        }) for i in range(5)])
+
+        # Still want 5 individual puts, but to 2 tables
+        self.assertEqual(blk._count, 2)
+        self.assertEqual(put_func.call_count, 5)
+
+    @patch('boto.dynamodb2.table.BatchTable.put_item')
+    @patch('boto.dynamodb2.table.Table.count')
+    @patch(DynamoDB.__module__ + '.connect_to_region')
+    def test_no_save_invalid(self, connect_func, count_func, put_func):
+        """ Make sure we only save valid signals """
+        blk = SaveCounterDynamoDB()
+        self.configure_block(blk, {
+            'log_level': 'DEBUG',
+            'hash_key': 'hash'
+        })
+        blk.start()
+
+        blk.process_signals([
+            Signal({'hash': 'hash val'}),
+            Signal({'nohash': 'nohash val'})])
+
+        # One table call, one real put call (one should have issued warning)
+        self.assertEqual(blk._count, 1)
+        self.assertEqual(put_func.call_count, 1)
+
+    @patch('boto.dynamodb2.table.BatchTable.put_item')
+    @patch('boto.dynamodb2.table.Table.count')
+    @patch(DynamoDB.__module__ + '.Table.create')
+    @patch(DynamoDB.__module__ + '.connect_to_region')
+    def test_table_lock(self, connect_func, create_func, count_func, put_func):
+        """ Make sure that if a table is creating it locks """
+        # We should return the error that the table is not found.
+        # We should only see this error returned once though, subsequent calls
+        # should use the cached version of the table.
+        count_func.side_effect = [JSONResponseError(
+            400,
+            "{'message': 'Requested resource not found: Table: T notfound', "
+            "'__type': 'com.amazonaws.dynamodb.v20120810"
+            "#ResourceNotFoundException'}")]
+        blk = SaveCounterDynamoDB()
+        self.configure_block(blk, {
+            'log_level': 'DEBUG',
+            'hash_key': 'hash'
+        })
+        # Simulate a table that is creating for a while
+        blk._get_table_status = MagicMock(
+            side_effect=['CREATING', 'CREATING', 'ACTIVE'])
+
+        # Send two threads to process signals at once
+        spawn(blk.process_signals, [Signal({'hash': 'value1'})])
+        spawn(blk.process_signals, [Signal({'hash': 'value2'})])
+
+        # We shouldn't have any save calls until the table creates.
+        # That takes 1.5 seconds, so let's see what we have after some of that.
+        sleep(0.7)
+        self.assertEqual(blk._count, 0)
+
+        # Give the table time to create...
+        sleep(1)
+
+        # Ok, it's created, we should see both signals get saved
+        self.assertEqual(blk._count, 2)
+
```

Comment:
Assert that create_func is only called once.

I know that this functionality is pretty much covered in test_no_table, but this assertion is relevant here too because of the presence of the lock.